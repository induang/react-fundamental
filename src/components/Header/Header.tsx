import { useLocation, useNavigate } from 'react-router-dom';
import { Box } from '@mui/system';
import { Grid, Button } from '@mui/material';
import { useAppSelector, useAppDispatch } from '../../redux/store/storeHook';

import { RootState } from '../../types';
import {
	authorizeThunk,
	logoutThunk,
	UserState,
	prepareForLogout,
} from '../../redux/slices/userSlice';

import LOGO from './components/Logo/Logo';
import { useEffect } from 'react';

function Header() {
	const navigate = useNavigate();
	const { pathname } = useLocation();
	const user = useAppSelector((state: RootState) => state.user);
	const dispatch = useAppDispatch();

	const handleLOGOUTClick = () => {
		dispatch(prepareForLogout());
		dispatch(logoutThunk());
		navigate('/login');
	};

	useEffect(() => {
		if (user.isAuth) dispatch(authorizeThunk());
	}, [user.isAuth]);

	return (
		<Box m={1}>
			<Grid
				container
				sx={{
					display: 'flex',
					justifyContent: 'space-between',
				}}
			>
				<Grid item>
					<LOGO />
				</Grid>
				<Grid item>
					<span style={{ fontSize: '20px' }} data-testid='test'>
						{user.isAuth && pathname !== '/registration' ? user.name : ' '}
					</span>
					<LogoutButton
						handleLOGOUTClick={handleLOGOUTClick}
						user={user}
						pathname={pathname}
					/>
				</Grid>
			</Grid>
		</Box>
	);
}
interface LogoutButtonProps {
	handleLOGOUTClick: () => void;
	user: UserState;
	pathname: string;
}

const LogoutButton = ({
	handleLOGOUTClick,
	user,
	pathname,
}: LogoutButtonProps) => {
	if (!user.isAuth || pathname === '/registration') {
		return <></>;
	} else {
		return (
			<Button
				variant='outlined'
				color='secondary'
				sx={{ ml: '20px' }}
				onClick={handleLOGOUTClick}
			>
				LOGOUT
			</Button>
		);
	}
};

export default Header;
