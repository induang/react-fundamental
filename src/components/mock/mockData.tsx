export const mockedCourse = {
	id: 'testID',
	title: 'testTitle',
	description: 'testDescrip',
	creationDate: '09/10/2022',
	duration: 12,
	authors: ['testAuthorId_1', 'testAuthorId_2'],
};

export const mockedAuthor_1 = {
	id: 'testAuthorId_1',
	name: 'testAuthorName_1',
};

export const mockedAuthor_2 = {
	id: 'testAuthorId_2',
	name: 'testAuthorName_2',
};

export const mockedUser = {
	isAuth: true,
	name: 'TestName',
	email: 'test',
	token: 'test',
	role: 'test',
};
