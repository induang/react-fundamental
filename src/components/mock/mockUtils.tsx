import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import { ToolkitStore } from '@reduxjs/toolkit/dist/configureStore';

import { RootState } from '../../redux/store/store';
import userSlice from '../../redux/slices/userSlice';
import courseSlice from '../../redux/slices/courseSlice';
import authorsSlice from '../../redux/slices/authorsSlice';

import {
	mockedUser,
	mockedCourse,
	mockedAuthor_1,
	mockedAuthor_2,
} from './mockData';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: ToolkitStore;
}
export function renderWithProviders(
	ui: React.ReactElement,
	{
		preloadedState = {
			user: mockedUser,
			courses: {
				courses: [mockedCourse],
			},
			authors: {
				authors: [mockedAuthor_1, mockedAuthor_2],
			},
		},
		store = configureStore({
			reducer: { user: userSlice, courses: courseSlice, authors: authorsSlice },
			preloadedState,
		}),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
		return <Provider store={store}>{children}</Provider>;
	}
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
