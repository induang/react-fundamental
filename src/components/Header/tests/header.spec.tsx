import { describe, it } from '@jest/globals';
import Header from '../Header';
import { BrowserRouter as Router } from 'react-router-dom';
import { renderWithProviders } from '../../mock/mockUtils';
import '@testing-library/jest-dom';

window.alert = jest.fn(() => ({}));

jest.mock('../../../services', () => {
	return {
		async fetchAuthorization() {
			return {
				result: {
					isAuth: true,
					name: 'testName',
					email: 'testName@email.com',
					role: 'admin',
				},
			};
		},
	};
});

describe('render header', () => {
	it('render correctly', async () => {
		const { container } = renderWithProviders(
			<Router>
				<Header />
			</Router>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('show Logo img', () => {
		const { queryByAltText } = renderWithProviders(
			<Router>
				<Header />
			</Router>
		);
		expect(queryByAltText('logo')).toBeTruthy();
	});

	it('show user name', () => {
		const { queryAllByRole } = renderWithProviders(
			<Router>
				<Header />
			</Router>
		);
		expect(queryAllByRole('generic')[0]).toHaveTextContent('TestName');
	});
});
