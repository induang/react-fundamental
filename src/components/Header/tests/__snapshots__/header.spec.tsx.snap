// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render header render correctly 1`] = `
<div
  class="MuiBox-root css-1ynyhby"
>
  <div
    class="MuiGrid-root MuiGrid-container css-1bukg3w-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <img
        alt="logo"
        class="w-24"
        src="logo.jpg"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <span
        data-testid="test"
        style="font-size: 20px;"
      >
        TestName
      </span>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-93jzzp-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        LOGOUT
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
